wifi: enterprise: Add support for runtime certificates

krish2718 opened the pull request:
Implements https://github.com/zephyrproject-rtos/zephyr/issues/79564 

* Move enterprise configs to a snippet
* Add TLS credentials configs to the new snippet
* Implement runtime certificate management using TLS credentials module
* Fix a bug in TLS credentials shell

The certificate installation is handled by a helper script which simplifies installation process, but even without that PR `cred` shell can be used manually but it's cumbersome.

krish2718 committed:
net: lib: tls_credentials_shell: Add a config for heap

The volatile backend stores the credentials on the heap, so, explicitly
add a config option that can be overridden in case there are more certs
than the default.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
wifi: Move Wi-Fi enterprise configs to a snippet

Instead of having an overlay move the Enterprise configurations to a
dedicated snippet so that it can be enabled with any sample.

Can be used along with Wi-Fi snippet e.g., `-S
"wifi-ipv4;wifi-enterprise"`.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
snippets: wifi-enterprise: Add support for TLS credentials

Enable TLS credentials shell to manager Wi-Fi enterprise certs.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: l2: wifi: Add support for run-time certificates

Using TLS credentials library add support for run-time certificates where
the installed certs are retrieved from the credential store (as of now
only volatile backend is tested).

This helps in production environments.

Implements #79564.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: lib: tls_credentials_shell: Fix crash during delete

Deletion of credential should use the pointer from the reference slot
not the temporary buffer, this causes a crash (unknown error).

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 commented:
@MaochenWang1 I have not done any AP mode testing on this as nRF70 doesn't support AP + Enterprise mode, it would e good if NXP can do some tests.

krish2718 committed:
net: l2: wifi: Make build time certs as optional

Certificates usage depends on STA/AP mode, but we don't have that
information at a build time, so, make all certs as optional and if a
file isn't found then generate an empty header so that corresponding C
code will be built.

Any missing mandatory certificates will be validated before connection
and connection is failed.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
samples: net: wifi: Add two variants of certs

RSA3K based certs are not supported on all platforms, so, keep both
variants, rsa2k (the older certs but with longer expiry 9999 days) and
rsa3k (latest ones) and we can have more variants in this folders.

Also, add a cmake variable to override the path with default as rsa3k.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

MaochenWang1 commented:
> @MaochenWang1 I have not done any AP mode testing on this as nRF70 doesn't support AP + Enterprise mode, it would e good if NXP can do some tests.

Sure, will review, test and feedback

krish2718 commented:
> > @MaochenWang1 I have not done any AP mode testing on this as nRF70 doesn't support AP + Enterprise mode, it would e good if NXP can do some tests.
> 
> Sure, will review, test and feedback

FYI, I am seeing issues even with build time certs, debugging now will push fixes soon.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -30,7 +30,29 @@ LOG_MODULE_REGISTER(net_wifi_shell, LOG_LEVEL_INF);
+	WIFI_CERT_CLIENT_P2_SEC_TAG,
+};
+
+typedef struct wifi_cert_data {
```
We try to avoid having typedef of structs and normally use those only for function pointers.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -121,27 +150,179 @@ static struct wifi_ap_sta_node sta_list[CONFIG_WIFI_SHELL_MAX_AP_STA];
 {
-	struct wifi_enterprise_creds_params params = {0};
+	wifi_cert_data_t certs_common[] = {
+		{TLS_CREDENTIAL_CA_CERTIFICATE, WIFI_CERT_CA_SEC_TAG, &params->ca_cert,
```
I would prefer that we use named struct fields when setting the values. Using positional data is error prone.

jukkar commented on subsys/net/lib/tls_credentials/tls_credentials_shell.c:
```diff
@@ -564,8 +564,8 @@ static int tls_cred_cmd_del(const struct shell *sh, size_t argc, char *argv[])
 		/* This was a credential we copied to heap. Clear and free it. */
-		memset((void *)cred_buf, 0, cred->len);
-		k_free((void *)cred_buf);
+		memset(cred_refs[ref_slot], 0, cred->len);
```
This seems to work but correct way to get the address of array item is `memset(&cred_refs[ref_slot], ...`

jukkar commented on subsys/net/lib/tls_credentials/tls_credentials_shell.c:
```diff
@@ -564,8 +564,8 @@ static int tls_cred_cmd_del(const struct shell *sh, size_t argc, char *argv[])
-		memset((void *)cred_buf, 0, cred->len);
-		k_free((void *)cred_buf);
+		memset(cred_refs[ref_slot], 0, cred->len);
+		k_free((void *)cred_refs[ref_slot]);
```
same here

jukkar commented on subsys/net/l2/wifi/CMakeLists.txt:
```diff
@@ -33,53 +33,28 @@ if(CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE AND CONFIG_NET_L2_WIFI_SHELL)
   # convert .pem files to array data at build time
```
For the commit `l2: wifi: ...`, please add `net: ` prefix

jukkar commented on samples/net/wifi/test_certs/install_certs.sh:
```diff
@@ -0,0 +1,124 @@
+#
+# Dependencies:
+#   - `device_credentials_installer` tool must be installed. You can install it using:
+#     pip3 install nrfcloud-utils
```
Does this mean that the instructions for installing the certs are only available for Nordic devices?

krish2718 replied:
No, that tools works with non-Nordic devices too, verified with STM32. BTW, this script should be part of the other PR will remove from here.

krish2718 committed:
doc: net: wifi: Fix the connect command

The command should work with existing certs rather than a generic
example, also fix the key-management.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 commented:
I resolve the compliance except for below: I tried to get checkpatch ignore using `--ignore FILE_PATH=a/samples/net/wifi/test_certs/` but looking at the perl script there is no such provision, it is trying to find typo in cert data :)

```
Running Checkpatch       tests in /home/tach/work/zephyr/zephyr ...
1 checks failed
ERROR   : Test Checkpatch failed: 
TYPO_SPELLING: 'jUs' may be misspelled - perhaps 'just'?
File:samples/net/wifi/test_certs/rsa2k/server.pem
Line:26

Complete results in compliance.xml
```

krish2718 commented:
> > > @MaochenWang1 I have not done any AP mode testing on this as nRF70 doesn't support AP + Enterprise mode, it would e good if NXP can do some tests.
> > 
> > 
> > Sure, will review, test and feedback
> 
> FYI, I am seeing issues even with build time certs, debugging now will push fixes soon.

Tested end-end, it is working fine.

kartben commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -30,37 +30,55 @@ Wi-Fi PSA crypto supported build
+    $ cp ca2.pem $CERTS_DIR
+    $ west build -p -b <board> samples/net/wifi -S wifi-enterprise
+
+or alternatively copy `rsa2k` certificates by changing the `CERTS_DIR` environment variable.
```
```suggestion
or alternatively copy ``rsa2k`` certificates by changing the `CERTS_DIR` environment variable.
```

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -121,27 +150,211 @@ static struct wifi_ap_sta_node sta_list[CONFIG_WIFI_SHELL_MAX_AP_STA];
+	memset(params, 0, sizeof(*params));
+}
+#else
+static inline void set_enterprise_creds_params(struct wifi_enterprise_creds_params *params,
```
BTW, I would suggest removing the inline from the various functions you are adding. Compiler is able to optimize the usage without the keyword anyway.

krish2718 replied:
Sure, but that would depend on optimizations levels and compiler, no? This is a standard practice to ensure it's uniform?

krish2718 replied:
Anyways, removed the inline keyword

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -121,27 +150,211 @@ static struct wifi_ap_sta_node sta_list[CONFIG_WIFI_SHELL_MAX_AP_STA];
+}
+#else
+static inline void set_enterprise_creds_params(struct wifi_enterprise_creds_params *params,
+				    bool is_ap)
```
Indentation is wrong

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1009,6 +1222,12 @@ static int cmd_wifi_disconnect(const struct shell *sh, size_t argc,
 
+#ifdef CONFIG_WIFI_SHELL_RUNTIME_CERTIFICATES
+	/* Clear the certificates */
+	clear_enterprise_creds_params(
+		&context.enterprise_creds_params);
```
Could be written to one line

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2010,6 +2229,13 @@ static int cmd_wifi_ap_disable(const struct shell *sh, size_t argc,
+
+#ifdef CONFIG_WIFI_SHELL_RUNTIME_CERTIFICATES
+	/* Clear the certificates */
+	clear_enterprise_creds_params(
+		&context.enterprise_creds_params);
```
Ditto

MaochenWang1 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -30,37 +30,63 @@ Wi-Fi PSA crypto supported build
+    $ cp ca2.pem $CERTS_DIR
+    $ west build -p -b <board> samples/net/wifi -S wifi-enterprise
+
+or alternatively copy ``rsa2k`` certificates by changing the ``CERTS_DIR`` environment variable.
```
when should use `rsa3k` or `rsa2k`? what's the difference？

krish2718 replied:
No guidelines, it's completely up to the user, just that the underlying board should support the algos.

MaochenWang1 commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -125,3 +125,23 @@ config WIFI_ENT_IDENTITY_MAX_USERS
+config HEAP_MEM_POOL_ADD_SIZE_WIFI_CERT
+	int "Wi-Fi enterprise security certificates memory pool size"
+	# STA - 6 certs and each assume 1500 bytes
+	default 12000
```
should `depend on WIFI_SHELL_RUNTIME_CERTIFICATES`, right?

MaochenWang1 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -121,27 +150,211 @@ static struct wifi_ap_sta_node sta_list[CONFIG_WIFI_SHELL_MAX_AP_STA];
+	const uint8_t *certs[] = {
+		params->ca_cert,
+		params->client_key,
+		params->server_cert,
```
`server_key` is missed

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -121,27 +150,211 @@ static struct wifi_ap_sta_node sta_list[CONFIG_WIFI_SHELL_MAX_AP_STA];
+			params->client_cert_len = ARRAY_SIZE(client_cert_test);
+			params->client_key = (uint8_t *)client_key_test;
+			params->client_key_len = ARRAY_SIZE(client_key_test);
+			params->ca_cert2 = (uint8_t *)ca_cert2_test;
```
No, I don't see it in https://w1.fi/cgit/hostap/plain/hostapd/hostapd.conf it's specifically for phase2 STA.

MaochenWang1 replied:
yes, you're right

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -121,27 +150,211 @@ static struct wifi_ap_sta_node sta_list[CONFIG_WIFI_SHELL_MAX_AP_STA];
+		{
+			.type = TLS_CREDENTIAL_CA_CERTIFICATE,
+			.sec_tag = WIFI_CERT_CA_P2_SEC_TAG,
+			.data = &params->ca_cert2,
```
As discussed above, it's STA only.

nordicjm commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -125,3 +125,26 @@ config WIFI_ENT_IDENTITY_MAX_USERS
+	  Uses the TLS credentials subsystem to store and manage the certificates.
+
+if WIFI_SHELL_RUNTIME_CERTIFICATES
+config HEAP_MEM_POOL_ADD_SIZE_WIFI_CERT
```
newline gap

nordicjm commented on subsys/net/lib/tls_credentials/Kconfig.shell:
```diff
@@ -32,4 +32,13 @@ config TLS_CREDENTIALS_SHELL_DIGEST_BUF_SIZE
 	   Also used to print error messages if digest generation fails.
 
+if TLS_CREDENTIALS_BACKEND_VOLATILE
+config HEAP_MEM_POOL_ADD_SIZE_TLS_CRED_SHELL
+	int "TLS credentials shell memory pool size"
+	# default 4 certs and each assume 1500 bytes
+	default 6000
+	help
+	  The size of the memory pool used by the TLS credentials shell.
```
newline gaps needed after if/before endif

nordicjm reviewed (commented):
Build system/snippet changes OK except nit.
Install certs script does not work on windows so a bit meh but not asking to recode it

krish2718 commented:
> Install certs script does not work on windows so a bit meh but not asking to recode it

Well, I haven't thought about anything else than Linux :). But let met think and see how to handle this in Windows and subbmit improvements later.

krish2718 commented:
Need to fix a build error and also add a test case to catch build errors and convert the script from shell->python for cross-platform support.

krish2718 committed:
tests: net: wifi: Add a test for Enterprise with run-time certs

This is needed to ensure run-time certs feature builds.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

nordicjm reviewed (dismissed):
Build system changes OK

kartben commented on samples/net/wifi/test_certs/install_certs.py:
```diff
@@ -0,0 +1,130 @@
+
+def check_requirements():
+    try:
+        installed_version = version('nrfcloud-utils')
```
Why is a generic wifi sample pulling in nrf cloud stuff? This sounds wrong.

krish2718 replied:
Though it's named `nrfcloud` it does work with non-nRF boards too, I have tested with STM32 Nucleo board. The reason for using `nrfcloud` util is it already supports TLS credential shell based device access. In the future we can write up a standalone tool part of Zephyr repo itself.

krish2718 replied:
As discussed on Discord, removed this commit from this PR, will raise a new PR with a standalone script (part of Zephyr itself).

kartben commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -30,37 +30,63 @@ Wi-Fi PSA crypto supported build
 
-For using variable size network buffer, the following overlay file can be used:
+To facilitate installation of the certificates, a helper script is provided in the ``samples/net/wifi/test_certs`` directory. The script can be used to install the certificates at runtime.
 
 .. code-block:: bash
 
-    $ west build -p -b <board> samples/net/wifi -- -DEXTRA_CONF_FILE=overlay-enterprise-variable-bufs.conf
+    $ samples/net/wifi/test_certs/install_certs.py -p samples/net/wifi/test_certs/rsa2k
 
+The script will install the certificates in the ``rsa2k`` directory to the TLS credentials store in the device over UART and using TLS credentials shell commands.
```
so all devices are supporting the AT commands this tool seems to be using?

krish2718 replied:
This doesn't use AT command, but TLS credentials shell (part of Zephyr)

kartben commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -30,37 +30,63 @@ Wi-Fi PSA crypto supported build
+Compile time certificates
+-------------------------
+
+Test certificates in PEM format are committed to the repo at :zephyr_file:`samples/net/wifi/test_certs` and the during the
```
really not a fan of the proliferation of binary certificates in-tree. Can't we just have instructions on what steps people sjhould take to generate them? It seems to me as this would be much more useful, too.

krish2718 replied:
Sure, but for a quick testing it's useful to have a `golden certs` that just work. Else any mistakes in cert generation are tough to debug.

kartben replied:
> Sure, but for a quick testing it's useful to have a `golden certs` that just work. Else any mistakes in cert generation are tough to debug.

Sorry but I still don't get why it's not possible to generate the certificates on the fly? Should be a few dozen lines of code using Python's `cryptography` package, I think? And this would have the merit to actually provide more guidance to the end user.

Also, where are these certificates coming from and are we even allowed to redistribute them (I wouldn't be surprised if they are part of some kind of Wi-Fi Alliance certification suite and not meant to be accessed by non-members ...)?

```console
$ openssl x509 -in client.pem -text -noout

Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            97:d4:07:ec:a6:05:15:13
        Signature Algorithm: sha384WithRSAEncryption
        Issuer: C=US, L=Santa Clara, O=Wi-Fi Alliance, CN=Suite B RSA 3k Root CA
...
```

krish2718 replied:
> Also, where are these certificates coming from and are we even allowed to redistribute them (I wouldn't be surprised if they are part of some kind of Wi-Fi Alliance certification suite and not meant to be accessed by non-members ...)?

Nope, please see the link above, it was generated by me using freeradius raddb scripts.

krish2718 replied:
> Sorry but I still don't get why it's not possible to generate the certificates on the fly? Should be a few dozen lines of code using Python's cryptography package, I think? And this would have the merit to actually provide more guidance to the end user.

It is possible, just that based on past experience with dealing with certs and unsupported ciphers, expired certs, mismatched certs, IMHO good to have a golden certs (compatible with the code) that just work. We can definitely do

* generation at runtime
* enforce user to provide certs

jukkar replied:
@kartben could you create an enhancement req for the built time cert generation, it is easier to track things there than discussing it here

rlubos replied:
What about root CA certs? We can't generate them, they must come from CA itself. Do we want to force users to download them on the fly when building sample?

kartben replied:
> What about root CA certs? We can't generate them, they must come from CA itself. Do we want to force users to download them on the fly when building sample?

It's already a scenario where user is "pretending" to be its own CA anyway, no? so why can't the CA be generated on-the-fly too?

```
openssl x509 -in rsa2k/ca.pem -text -noout

Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            1f:ad:78:ce:f9:82:9e:0a:52:73:6e:81:2c:b1:6e:da:5a:2a:bd:f8
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=FR, ST=Radius, L=Somewhere, O=Example Inc, emailAddress=admin@example.org, CN=Example Certificate Authority
```

krish2718 replied:
> why do they appear to be issued by Wi-Fi Alliance then?

Sorry, I am missing something. Where is Wi-Fi alliance coming from? The issuer is `Example certificate authority`, no?

kartben replied:
> > why do they appear to be issued by Wi-Fi Alliance then?
> 
> Sorry, I am missing something. Where is Wi-Fi alliance coming from? The issuer is `Example certificate authority`, no?

Not for the rsa3k certs, no, and I have concerns these can't be redistributed outside of Wi-Fi Alliance certification program. This does need to be clarified.

krish2718 replied:
Ah thanks, I understand the confusion now, I was looking the ones that I have added in this PR. The WFA ones are existing ones already in `main` I have just moved them to `rsa3k` directory, they are submitted by NXP. + @MaochenWang1 @fengming-ye can you please respond to Ben's question about redistributing those certs in Zephyr?

jukkar replied:
@kartben the Wi-Fi alliance certs are already in main and this PR is just moving them around. I suggest that we deal the cert distribution issue and possible removal from zephyr main in a separate PR.

carlescufi replied:
@jukka could you please create a tracking issue for this redistribution license problem, so we ensure that this is resolved ASAP?

jukkar replied:
> create a tracking issue for this redistribution license problem, so we ensure that this is resolved ASAP?

Sure, it is here #88771
